library/src/main/scala/scala/scalajs/js/annotation/internal/JSOptional.scala

```diff
@@ -0,0 +1,19 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2016, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package scala.scalajs.js.annotation.internal
+
+import scala.annotation.meta._
+
+/** IMPLEMENTATION DETAIL: Marks a member of a JavaScript trait as an optional
+ *  field, that must be emitted.
+ *
+ *  Do not use this annotation yourself.
+ */
+@field @getter @setter
+class JSOptional extends scala.annotation.StaticAnnotation
```

Comment:
The annotation is very important. In fact in this version we only truly test the annotation on *mixed in* fields and defs. We are using the fact that `mixin` copies annotations to the fields and methods it creates to be able to identify those created from JSOptional fields/defs in traits.